hs/src/Reach/AST/DLBase.hs

```diff
@@ -45,6 +69,45 @@ data DLVar = DLVar SrcLoc String DLType Int
 instance Eq DLVar where
   (DLVar _ _ _ x) == (DLVar _ _ _ y) = x == y
 
+-- XXX better error message for stuff like Array<Fun> or Array<Forall>
+-- that can't exist in DL-land
+st2dt :: HasCallStack => SLType -> DLType
+st2dt = \case
+  ST_Null -> T_Null
+  ST_Bool -> T_Bool
+  ST_UInt -> T_UInt
+  ST_Bytes i -> T_Bytes i
+  ST_Digest -> T_Digest
+  ST_Address -> T_Address
+  ST_Array ty i -> T_Array (st2dt ty) i
+  ST_Tuple tys -> T_Tuple (map st2dt tys)
+  ST_Object tyMap -> T_Object (M.map st2dt tyMap)
+  ST_Data tyMap -> T_Data (M.map st2dt tyMap)
+  -- XXX consider using Maybe so that callers have to handle the error case
+  t@(ST_Fun {}) -> error $ "ST_Fun not a dt: " <> show t
+  t@(ST_Forall {}) -> error $ "ST_Forall not a dt: " <> show t
+  t@(ST_Var {}) -> error $ "ST_Var not a dt: " <> show t
+  t@(ST_Type {}) -> error $ "ST_Type not a dt: " <> show t
```

Comment:
I try not to use `error`, because I want us to think that `impossible` means that some earlier part of the compiler has not done its job to enforce some invariant. In this case, I believe (?) that `Type.hs` is responsible for this kind of logic and either it would enforce it (so this is `impossible`) or it is a consumer of this, so it needs it to be a `Maybe` so that it can blame the user